core/priority/prioritiser.go

```diff
@@ -0,0 +1,279 @@
+type msgProvider func(slot int64) (*pbv1.PriorityMsg, error)
+
+// msgValidator abstracts validation of a received priority messages.
+type msgValidator func(*pbv1.PriorityMsg) error
+
+// tickerProvider abstracts the consensus timeout ticker (for testing purposes only).
+type tickerProvider func() (<-chan time.Time, func())
+
+// subscriber abstracts the output subscriber callbacks of Prioritiser.
+type subscriber func(ctx context.Context, slot int64, topic string, priorities []string) error
+
+// NewForT returns a new prioritiser for testing.
+func NewForT(_ *testing.T, transport Transport,
+	consensus Consensus, msgProvider msgProvider, msgValidator msgValidator,
+	consensusTimeout time.Duration, tickerProvider tickerProvider,
+) *Prioritiser {
+	n := &Prioritiser{
+		transport:        transport,
+		consensus:        consensus,
+		msgProvider:      msgProvider,
+		msgValidator:     msgValidator,
+		consensusTimeout: consensusTimeout,
+		tickerProvider:   tickerProvider,
+		subs:             make(map[string][]subscriber),
+		trigger:          make(chan int64, 1), // Buffer a single trigger
+	}
+
+	// Wire consensus output to Prioritiser subscribers.
+	consensus.Subscribe(func(ctx context.Context, slot int64, result *pbv1.PriorityResult) error {
+		for _, topic := range result.Topics {
+			for _, sub := range n.subs[topic.Topic] {
+				if err := sub(ctx, slot, topic.Topic, topic.Priorities); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+
+	// Register prioritiser protocol handler.
+	transport.ReceiveSend(
+		func() proto.Message { return new(pbv1.PriorityMsg) },
+		func(ctx context.Context, pID peer.ID, msg proto.Message) (proto.Message, error) {
+			prioMsg, ok := msg.(*pbv1.PriorityMsg)
+			if !ok {
+				return nil, errors.New("invalid priority message")
+			}
+
+			return n.handleRequest(ctx, pID, prioMsg)
+		},
+		protocolID)
+
+	return n
+}
+
+// Prioritiser resolves cluster wide priorities.
+type Prioritiser struct {
+	quit             chan struct{}
+	trigger          chan int64
+	received         chan *pbv1.PriorityMsg
+	minRequired      int
+	consensusTimeout time.Duration
+	transport        Transport
+	peers            []peer.ID
+	consensus        Consensus
+	msgProvider      msgProvider
+	msgValidator     msgValidator
+	tickerProvider   tickerProvider
+	subs             map[string][]subscriber
+}
+
+// Subscribe registers a prioritiser output subscriber function.
+// This is not thread safe and MUST NOT be called after Run.
+func (n *Prioritiser) Subscribe(topic string, fn subscriber) {
+	n.subs[topic] = append(n.subs[topic], fn)
+}
+
+// Prioritise starts a new prioritisation round for the provided slot.
+func (n *Prioritiser) Prioritise(slot int64) {
+	select {
+	case n.trigger <- slot:
+	case <-n.quit:
+	}
+}
+
+// Run runs the prioritiser until the context is cancelled.
+// Note this will panic if called multiple times.
+//nolint:gocognit // Not that bad I feel.
```

Comment:
`startConsensus` can be safely written outside `Run`. I think it's independent and doesn't have strong benefits?

Reply:
not so sure about that